asynciterator.ts

```diff
@@ -1252,6 +1254,135 @@ function destinationFillBuffer<S>(this: InternalSource<S>) {
 }
 
 
+export abstract class SynchronousTransformIterator<S, D = S> extends AsyncIterator<D> {
+  private _destroySource: boolean;
+  protected constructor(protected _source: AsyncIterator<S>, protected options: { destroySource?: boolean } = {}) {
+    /* eslint-disable no-use-before-define */
+    super();
+    this._destroySource = options.destroySource !== false;
+    const cleanup = () => {
+      _source.removeListener('end', onSourceEnd);
+      _source.removeListener('error', onSourceError);
+      _source.removeListener('readable', onSourceReadable);
+    };
+    const onSourceEnd = () => {
+      cleanup();
+      this.close();
+    };
+    const onSourceError = (err: Error) => {
+      this.emit('error', err);
+    };
+    const onSourceReadable = () => {
+      if (this.readable) {
+        // TODO: I'm not completely sure as to why this is needed but without
+        //       the following line, some use cases relying on flow mode (i.e.
+        //       consuming items via `on('data', (data) => {})`) do not work.
+        //       It looks like the debouncing that happens in `set readable()`
+        //       in `AsyncIterator` prevents the event from firing as `this`
+        //       is already readable.
+        this.emit('readable');
+      }
+      else {
+        this.readable = true;
+      }
+    };
+    _source.on('end', onSourceEnd);
+    _source.on('error', onSourceError);
+    _source.on('readable', onSourceReadable);
+    if (_source.done)
+      onSourceEnd();
+    else if (_source.readable)
+      onSourceReadable();
+  }
+
+  destroy(cause?: Error): void {
+    this._source.destroy(cause);
+    super.destroy(cause);
+  }
+
+  public close() {
+    if (this._destroySource)
+      this._source.destroy();
+    super.close();
+  }
+}
+
+interface Transform {
+  fn: Function,
+  next?: Transform
+}
+
+export class SyncTransformIterator<T, D = T> extends SynchronousTransformIterator<T, D> {
+  private _funcs?: Function[];
+
+  constructor(private source: AsyncIterator<T>, private transforms: Transform, private upstream: AsyncIterator<any> = source, options?: { destroySource?: boolean }) {
+    // Subscribe the iterator directly upstream rather than the original source to avoid over-subscribing
+    // listeners to the original source
+    super(upstream, options);
+  }
+
+  get funcs() {
+    if (!this._funcs) {
+      this._funcs = [];
+      // eslint-disable-next-line prefer-destructuring
+      let transforms: Transform | undefined = this.transforms;
+      do
+        this._funcs.push(transforms.fn);
+      // eslint-disable-next-line no-cond-assign
+      while (transforms = transforms.next);
+    }
+    return this._funcs;
+  }
+
+  read(): D | null {
+    const { source, funcs } = this;
+    let item;
+    outer: while ((item = source.read()) !== null) {
+      // Do not use a for-of loop here, it slows down transformations
```

Comment:
What about a `reduceRight` loop, which is what this seems to be?
But I see we have an early stop condition indeed.

Reply:
Yeah, I've tried to go down that path, but it is very necessary to have the break condition rather than just forwarding nulls as this can allow filtered operations to terminate much faster; that said, we can remove the labels by refactoring 1344-1347 into a separate function and use return whilst seeing similar perf. outcomes

Reply:
We can definitely do this with two regular loops. Quick cut 'n' paste draft:

```typescript
  read(): D | null {
    const { source, funcs } = this;
    let item;
    while ((item = source.read()) !== null) {
      let next = funcs.length;
      let transformed = item;
      while (transformed !== null) {
        // …
      }
      if (transformed !== null)
        return transformed;
    }
    return null;
  }
  ```

Reply:
I've pushed an alternative that I had worked on before; but happy to look into this one further if you think the approach I now have up is still confusing